xds/internal/balancer/cdsbalancer/cluster_handler_test.go

```diff
@@ -0,0 +1,390 @@
+func (s) TestUpdateRootClusterLogicalDNSSuccess(t *testing.T) {
+	ch, fakeClient := setupTests(t)
+	ch.updateRootCluster(logicalDNSService)
+
+	ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer ctxCancel()
+	gotCluster, err := fakeClient.WaitForWatchCluster(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
+	}
+	if gotCluster != logicalDNSService {
+		t.Fatalf("xdsClient.WatchCDS called for cluster: %v, want: %v", gotCluster, logicalDNSService)
+	}
+
+	fakeClient.InvokeWatchClusterCallback(xdsclient.ClusterUpdate{
+		ClusterType: xdsclient.ClusterTypeLogicalDNS,
+		ServiceName: logicalDNSService,
+	}, nil)
+
+	select {
+		case chu := <-ch.updateChannel:
+			if len(chu.chu) != 1 {
+				t.Fatal("Cluster Update passed to CDS should only have one update as type EDS")
+			}
+			if chu.chu[0].ClusterType != xdsclient.ClusterTypeLogicalDNS {
+				t.Fatalf("ClusterUpdate Type received: %v, ClusterUpdate wanted: %v", chu, xdsclient.ClusterTypeLogicalDNS)
+			}
+			if chu.chu[0].ServiceName != logicalDNSService {
+				t.Fatalf("ClusterUpdate ServiceName received: %v, ClusterUpdate wanted: %v", chu.chu[0].ServiceName, logicalDNSService)
+			}
+		case <-ctx.Done():
+			t.Fatal("Timed out waiting for update from updateChannel.")
+	}
+}
+
+func (s) TestUpdateRootClusterAggregateSuccess(t *testing.T) {
+	ch, fakeClient := setupTests(t)
+	ch.updateRootCluster(aggregateClusterService)
+
+	ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer ctxCancel()
+	gotCluster, err := fakeClient.WaitForWatchCluster(ctx)
+	if err != nil {
+		t.Fatalf("xdsClient.WatchCDS failed with error: %v", err)
+	}
+	if gotCluster != aggregateClusterService {
+		t.Fatalf("xdsClient.WatchCDS called for cluster: %v, want: %v", gotCluster, aggregateClusterService)
+	}
+	// The xdsClient telling the clusterNode that the cluster type is an aggregate cluster which will cause a lot of
+	// downstream behavior. For a cluster type that isn't an aggregate, the behavior is simple. The clusterNode will
+	// simply get a successful update, which will then ping the clusterHandler which will successfully build an update
+	// to send to the CDS policy. In the aggregate cluster case, the handleResp callback must also start watches for
+	// the aggregate cluster's children. The ping to the clusterHandler at the end of handleResp should be a no-op,
+	// as neither the EDS or LogicalDNS child clusters have received an update yet.
+	fakeClient.InvokeWatchClusterCallback(xdsclient.ClusterUpdate{
+		ClusterType: xdsclient.ClusterTypeAggregate,
+		ServiceName: aggregateClusterService,
+		PrioritizedClusterNames: []string{edsService, logicalDNSService},
+	}, nil)
+
+	// xds client should be called to start a watch for one of the child clusters of the aggregate, which is either an
+	// EDS Service or a LogicalDNS. The construction/iteration through the map in handleResp() is nondeterministic, so
```

Comment:
It should be deterministic.
See my other comment about map shuffling items.

Reply:
I changed it to an ordered list. Will switch this back to my check before non determinism, which was one after the next.

Reply:
Changed validations to have logical ordering.